### PR TITLE
added floating lantern box that spawns floating lanterns

### PR DESCRIPTION
**Pull request by MrRoboman**
Testing
Rez lantern box. 
It should rez four lanterns.

Near grab, far grab, or click and release a lantern.
Lantern will begin to float away.
When the lantern is removed from the box, a new lantern will be rezzed.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by cozza13** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1873558df15ec8d8bbbdcac7a275738469c1af10 available at: https://gist.github.com/e234c271a76dcebd2d12f86986a5df1e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-1873558.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-1873558.dmg

**Comment by cozza13** [comment]
ok to test

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1873558df15ec8d8bbbdcac7a275738469c1af10 available at: https://gist.github.com/e234c271a76dcebd2d12f86986a5df1e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-1873558.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-1873558.dmg

**Comment by Nex-Pro** [comment]
Testing this win10 and Vive

**Comment by Nex-Pro** [comment]
### FAIL 

ran scripts got nothing apart from some log output.


[05/09 16:11:36] [DEBUG] [hifi.scriptengine] Done downloading script at: "file:///E:/Program Files/High Fidelity - PR10402/scripts/tutorials/entity_scripts/floatingLanternBox.js"
[05/09 16:11:36] [DEBUG] [hifi.scriptengine] loadURL "file:///E:/Program Files/High Fidelity - PR10402/scripts/tutorials/entity_scripts/floatingLanternBox.js" "Success" QThread(0x1f62824fb90, name = "Main Thread")
[05/09 16:11:36] [INFO] [hifi.scriptengine] "Script Engine starting:floatingLanternBox.js"
[05/09 16:11:42] [DEBUG] [hifi.interface.deadlock] DEADLOCK WATCHDOG WARNING: lastHeartbeatAge: 271 elapsedMovingAverage: 818 maxElapsed: 9201248 PREVIOUS maxElapsedAverage: 815 NEW maxElapsedAverage: 818 ** NEW MAX ELAPSED AVERAGE ** samples: 121017

**Comment by MrRoboman** [comment]
Did you run all the scripts? You only need to run createFloatingLanternBox.js
If you run 
https://raw.githubusercontent.com/MrRoboman/hifi/1873558df15ec8d8bbbdcac7a275738469c1af10/scripts/tutorials/createFloatingLanternBox.js
That should build the floating lantern box and load the other two scripts.

**Comment by Nex-Pro** [comment]
Will try again. In future please  put this kind of info into the test plan.

**Comment by MrRoboman** [comment]
Will do in the future.

**Comment by Nex-Pro** [comment]
### PASS

@MrRoboman  This works fine.

Clean up works well.
Script seems robust.

Good work on making a robust script that resists my attempts to make things go bad.  You seem to have covered all the obvious stuff such as scooping the lanterns out with another object and capping the box with another entity. Well done.

Only thing I'd change is a small random Y angular velocity for each lantern and a little irregularity to the gravity ( these are lifted by flames ) to add a more life to the up moving stream of lanterns.

File: scripts/tutorials/createFloatingLanternBox.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+var COMPOUND_SHAPE_URL = "http://hifi-content.s3.amazonaws.com/Examples%20Content/production/maracas/woodenCrate_phys.obj";
+var MODEL_URL = "http://hifi-content.s3.amazonaws.com/Examples%20Content/production/maracas/woodenCrate_VR.fbx";
+var SCRIPT_URL = Script.resolvePath("./entity_scripts/floatingLanternBox.js?v=" + Date.now());
+var START_POSITION = Vec3.sum(MyAvatar.position, Vec3.multiply(Quat.getFront(MyAvatar.orientation), 2));
+START_POSITION.y -= .6;
+var LIFETIME = 3600;
+
+var lanternBox = {
+  type: "Model",
```
Comment (ZappoMan) [inline scripts/tutorials/createFloatingLanternBox.js]: indent

File: scripts/tutorials/entity_scripts/floatingLantern.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+// Created by MrRoboman on 17/05/04
+// Copyright 2017 High Fidelity, Inc.
+//
+// Makes floating lanterns rise upon being released and corrects their rotation as the fly.
+//
+// Distributed under the Apache License, Version 2.0.
+// See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+
+(function() {
+  var _this;
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLantern.js]: indent

File: scripts/tutorials/entity_scripts/floatingLantern.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          y: 0.5,
+          z: 0
+        }
+      });
+    },
+
+    update: function(dt) {
+      var lanternProps = Entities.getEntityProperties(_this.entityID);
+
+      if(lanternProps && lanternProps.rotation && lanternProps.owningAvatarID === MyAvatar.sessionUUID) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLantern.js]: space

File: scripts/tutorials/entity_scripts/floatingLantern.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+          Math.abs(lanternProps.angularVelocity.z) < SLOW_SPIN_THRESHOLD
+        );
+
+        var rotationComplete = (
+           Math.abs(lanternProps.rotation.x - HOME_ROTATION.x) < ROTATION_COMPLETE_THRESHOLD &&
+           Math.abs(lanternProps.rotation.y - HOME_ROTATION.y) < ROTATION_COMPLETE_THRESHOLD &&
+           Math.abs(lanternProps.rotation.z - HOME_ROTATION.z) < ROTATION_COMPLETE_THRESHOLD
+         );
+
+        if(spinningSlowly && !rotationComplete) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLantern.js]: space

File: scripts/tutorials/entity_scripts/floatingLantern.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+              y: 0,
+              z: 0
+            }
+          });
+        }
+      }
+    },
+
+    connectUpdate: function() {
+      if(!this.updateConnected) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLantern.js]: space

File: scripts/tutorials/entity_scripts/floatingLantern.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        this.updateConnected = true;
+        Script.update.connect(this.update);
+      }
+    },
+
+    disconnectUpdate: function() {
+      if(this.updateConnected) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLantern.js]: space

File: scripts/tutorials/entity_scripts/floatingLanternBox.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+// Copyright 2017 High Fidelity, Inc.
+//
+// Spawns new floating lanterns every couple seconds if the old ones have been removed.
+//
+// Distributed under the Apache License, Version 2.0.
+// See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+
+(function() {
+
+  var _this;
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLanternBox.js]: indent

File: scripts/tutorials/entity_scripts/floatingLanternBox.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    _this = this;
+  };
+
+  lanternBox.prototype = {
+
+    preload: function(entityID) {
+      this.entityID = entityID;
+      var props = Entities.getEntityProperties(this.entityID);
+
+      if(props.owningAvatarID === MyAvatar.sessionUUID){
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLanternBox.js]: space

File: scripts/tutorials/entity_scripts/floatingLanternBox.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+      if(this.respawnTimer)
+        Script.clearInterval(this.respawnTimer);
+    },
+
+    spawnAllLanterns: function() {
+      var props = Entities.getEntityProperties(this.entityID);
+      var lanternCount = 0;
+      var nearbyEntities = Entities.findEntities(props.position, props.dimensions.x * 0.75);
+
+      for(var i = 0; i < nearbyEntities.length; i++) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLanternBox.js]: space

File: scripts/tutorials/entity_scripts/floatingLanternBox.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        var name = Entities.getEntityProperties(nearbyEntities[i], ["name"]).name;
+        if(name === "Floating Lantern") {
+          lanternCount++;
+        }
+      }
+
+      while(lanternCount++ < MAX_LANTERNS) {
```
Comment (ZappoMan) [inline scripts/tutorials/entity_scripts/floatingLanternBox.js]: space

**Comment by MrRoboman** [comment]
Fixed coding standards

**Comment by hifi-gustavo** [comment]
Build console output for commit: a12e8e34dbd5833b8013cfbcfbeac267a32538e1 available at: https://gist.github.com/e234c271a76dcebd2d12f86986a5df1e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-a12e8e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-a12e8e3.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9fc0ad0d28dc79aa258e82f7c9af0229163570a6 available at: https://gist.github.com/e234c271a76dcebd2d12f86986a5df1e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-9fc0ad0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-9fc0ad0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f3d8d1641f62b36365671dcf8f085bbfc97aa40e available at: https://gist.github.com/e234c271a76dcebd2d12f86986a5df1e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-f3d8d16.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10402-f3d8d16.dmg

**Pull request merged**